src/WebJobs.Extensions.DocumentDB/DocumentDBUtility.cs

```diff
@@ -0,0 +1,136 @@
+                    throw;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the core DocumentDB type for the supplied parameter.
+        /// </summary>
+        /// <remarks>
+        /// For example, the core Type is T in the following parameters:
+        /// <list type="bullet">
+        /// <item><description><see cref="ICollector{T}"/></description></item>
+        /// <item><description>T[]</description></item>
+        /// <item><description>out T</description></item>
+        /// <item><description>out T[]</description></item>
+        /// </list>
+        /// </remarks>
+        /// <param name="type">The Type to evaluate.</param>
+        /// <returns>The core Type</returns>
+        public static Type GetCoreType(Type type)
+        {
+            Type coreType = type;
+            if (coreType.IsByRef)
+            {
+                coreType = coreType.GetElementType();
+            }
+
+            if (coreType.IsArray)
+            {
+                return coreType.GetElementType();
+            }
+
+            if (coreType.IsGenericType)
+            {
+                Type genericArgType = null;
+                if (TryGetSingleGenericArgument(coreType, out genericArgType))
+                {
+                    return genericArgType;
+                }
+
+                throw new InvalidOperationException("DocumentDB parameter types can only have one generic argument.");
+            }
+
+            return coreType;
+        }
+
+        /// <summary>
+        /// Checks whether the specified type has a single generic argument. If so,
+        /// that argument is returned via the out parameter.
+        /// </summary>
+        /// <param name="genericType">The generic type.</param>
+        /// <param name="genericArgumentType">The single generic argument.</param>
+        /// <returns>true if there was a single generic argument. Otherwise, false.</returns>
+        public static bool TryGetSingleGenericArgument(Type genericType, out Type genericArgumentType)
+        {
+            genericArgumentType = null;
+            Type[] genericArgTypes = genericType.GetGenericArguments();
+
+            if (genericArgTypes.Length != 1)
+            {
+                return false;
+            }
+
+            genericArgumentType = genericArgTypes[0];
+            return true;
+        }
+
+        /// <summary>
+        /// Execute the function with retries on throttle.
+        /// </summary>
+        /// <typeparam name="T">The type of return value from the execution.</typeparam>
+        /// <param name="function">The function to execute.</param>
+        /// <returns>The response from the execution.</returns>
+        // Taken from: https://github.com/ryancrawcour/azure-documentdb-dotnet/blob/e7dd2f685554b0a9def63c8925f8e3ef2ad3bff8/samples/code-samples/Shared/Util/DocumentClientHelper.cs
+        public static async Task<T> ExecuteWithRetriesAsync<T>(Func<Task<T>> function)
```

Comment:
Good question. I'm not sure I've seen that anywhere else. I'll ask.

---
In reply to: [56258775](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/47#discussion_r56258775) [](ancestors = 56258775)